_client/src/Network/ApiManager.cpp

```diff
@@ -52,13 +52,65 @@ void ApiManager::registerUser(const QString &username, const QString &password)
         dataset);
 }
 
+void ApiManager::getUsersRoles()
+{
+    SPDLOG_TRACE("ApiManager::getUsersRole");
+
+    m_networkService->sendRequest(
+        Api::Endpoints::Users::ROLES,
+        QNetworkAccessManager::Operation::GetOperation);
+}
+
+void ApiManager::setNewRole(const QString &new_role)
+{
+    SPDLOG_TRACE("ApiManager::setNewRole");
+
+    Dataset dataset;
+    dataset["new_role"] = {new_role};
+
+    m_networkService->sendRequest(
+        Api::Endpoints::Users::ROLES,
+        QNetworkAccessManager::Operation::PostOperation,
+        dataset);
+}
+
+void ApiManager::editRoleId(const QString &editrole, const QString &id)
+{
+    SPDLOG_TRACE("ApiManager::editRole");
+
+    Dataset dataset;
+    dataset["editrole"] = {editrole};
+
+    m_networkService->sendRequest(
+        Api::Endpoints::Users::ROLES + "/" + id,
+        QNetworkAccessManager::Operation::PutOperation,
+        dataset);
+}
+
+void ApiManager::deleteRole(const QString &id)
+{
+    SPDLOG_TRACE("ApiManager::deleteRole");
+
+    m_networkService->sendRequest(
+        Api::Endpoints::Users::ROLES + "/" + id,
+        QNetworkAccessManager::Operation::DeleteOperation);
+}
+
 void ApiManager::setupHandlers()
 {
     SPDLOG_TRACE("ApiManager::setupHandlers");
     m_responseHandlers[Api::Endpoints::Users::LOGIN] = [this](const Dataset &dataset)
     { handleLoginResponse(dataset); };
     m_responseHandlers[Api::Endpoints::Users::REGISTER] = [this](const Dataset &dataset)
     { handleRegistrationResponse(dataset); };
+    m_responseHandlers[Api::Endpoints::Users::ROLES] = [this](const Dataset &dataset)
```

Comment:
Bad code.

What do you expect this code to do? Think when you code.

This 4 `m_responseHandlers[Api::Endpoints::Users::ROLES] =` just overwriting each other.
In the end only the last one will call `handleRegistrationResponse`, what is wrong

Just wtf.
Figure out HOW this code work and WHY.
If you don't understand - ASK.
You made such shit and not working code and then asked for review... It's bad practice and nobody will like it.